stablehlo/transforms/optimization/StablehloAggressiveFolder.cpp

```diff
@@ -448,6 +448,59 @@ struct FoldCompareOpPattern : public ShapeOpRewritePattern<CompareOp> {
   };
 };
 
+//////////////////////////////////
+// CaseOp
+/////////////////////////////////
+
+class InlineCaseOpWithConstantBranchIndex
+    : public FoldOpRewritePattern<CaseOp> {
+ public:
+  using FoldOpRewritePattern::FoldOpRewritePattern;
+
+  LogicalResult matchAndRewrite(CaseOp op,
+                                PatternRewriter& rewriter) const override {
+    Value branchIndexArgument = op.getIndex();
+    DenseIntElementsAttr indexAttr;
+    if (!matchPattern(branchIndexArgument, m_Constant(&indexAttr)))
+      return rewriter.notifyMatchFailure(op, "Branch index is not a constant.");
+
+    int64_t selectedBranchIndex =
+        indexAttr.getSplatValue<IntegerAttr>().getValue().getSExtValue();
+    // If the branch index is OOB, the last branch is executed by default:
+    // https://openxla.org/xla/operation_semantics#conditional
+    if (selectedBranchIndex < 0 || selectedBranchIndex >= op.getNumRegions())
+      selectedBranchIndex = op.getNumRegions() - 1;
+
+    if (op->use_empty())
+      return rewriter.notifyMatchFailure(op, "The case op's result is unused.");
+
+    Region& region = op.getRegion(selectedBranchIndex);
+
+    if (!llvm::hasSingleElement(region))
```

Comment:
We can probably assert / omit this. I believe our ODS dictates this as well:

```
VariadicRegion<SizedRegion<1>>:$branches
```

Is there a CaseOp API we can use to simplify this and get the block directly? `op.getBranch` or something? If not then LGTM.